File: src_fortran/sub/adjustmodel.f
c
c
c
c
c
c
      subroutine ADJUSTMODEL(damp)
c
c     adjust model-vector by the solution just obtained
c
      implicit none
      real damp
      include '../inc/vel_com.inc'
c
      integer jj,i,n,iminold,j,itime,j1,j2,k,k2,j11,j22
      integer juliam
      integer kj,jjj,k1,ksta1,kk1,ksta2
      real cc(ist)
      real dmin
c
c     adjust hypocenters of this iteration and output them:
c
      jj=0
c
      do 29 i=1,legs
      n=4
      if(i.gt.neqs) n=1   ! for shots adjust only the origin-time !!!
ccc      call TIMECLEAR(iyr(i),imo(i),iday(i),ihr(i),imin(i),e(1,i),iminold)
      if(iyr(i).lt.100)then
         iminold=JULIAM(iyr(i)+1900,imo(i),iday(i),ihr(i),imin(i))
      else
         iminold=JULIAM(iyr(i),imo(i),iday(i),ihr(i),imin(i))
      endif
      write(*,'(A,I5,5F12.4)') 'DEBUG: i=',i,(e(j,i),j=1,4)
      do j=1,n
         jj=jj+1
         e(j,i)=e(j,i) + b(jj)
      enddo
      if(e(1,i).lt.0.0)then   ! change reference minute
        call TIMECLEAR(iyr(i),imo(i),iday(i),ihr(i),imin(i),e(1,i),
     &                                                      itime)
        do j=1,knobs(i)
           pt(j,i)=pt(j,i)+(iminold-itime)*60.
        enddo
      endif
      do j=1,3
         isconstrain(j)=0
      enddo
      iconstrain(i)=0
c
c     constrain focal depth if necessary
c
      if(isingle.ne.0)then
         if(nitt.lt.2)then
            isconstrain(1)=1
c            write(16,*)' *** nitt<2 --> ',
c     &                 'depth-adjustment := 0.0 for event ',i
            e(4,i)=e(4,i)-b(4)
            b(4)=0.0
         endif     
         if(igap(1).gt.250)then
            dmin=999.9   ! minimum distance (epicenter --> receiver)
            do k=1,knobs(1)
               delta=SQRT(  (e(2,1)-d(k,1,1))**2
     &                    + (e(3,1)-d(k,2,1))**2  )
               if(delta.lt.dmin) then
	          dmin=delta
	       endif
            enddo
	    if (dmin.gt.15.) then
               isconstrain(2)=1
               if(iconstrain(i).eq.1) iconstrain(i)=3
c              write(16,*)' *** igap>250 --> ',
c     &                  'depth-adjustment := 0.0 for event ',i
               e(4,i)=e(4,i)-b(4)
               b(4)=0.0
	    endif
         endif
         if(abs(b(4)).gt.zadj)then
            isconstrain(3)=1
c            write(16,*)' *** depth-adjustment constrained for event ',i
            if(b(4).gt.0.0)then
               e(4,i)=e(4,i)-b(4)+zadj
               b(4)=zadj
            else
               e(4,i)=e(4,i)-b(4)-zadj
               b(4)=-zadj
            endif
         endif
      else    ! simultanous inversion
         if(i.le.neqs.and.abs(b(jj)).gt.zadj)then ! b(jj)=depth-adj. if event i
            iconstrain(i)=1
c            write(16,*)' *** depth-adjustment constrained for event ',i
            if(b(jj).gt.0.0)then
               e(4,i)=e(4,i)-b(jj)+zadj
               b(jj)=zadj
            else
               e(4,i)=e(4,i)-b(jj)-zadj
               b(jj)=-zadj
            endif
         endif
      endif
c      effdeltaz(i)=0.0
      if(itopo.gt.0)then
         if(e(4,i).lt.0.0)then  ! depth above sea-level...so near surface !
            call CHTOP(-e(2,i),e(3,i),zmin,
     &                 topo1file,topo2file) ! zmin:==surface at this point
         else
            zmin=0.0 ! depth below zero ... so zmin:==0.0 is sufficiant !
         endif
      endif
      if(e(4,i).lt.zmin.or.(ifixsolution.gt.0.and.e(4,i).le.0.0))then
c         effdeltaz(i)=b(jj)-(e(4,i)-zmin)
c
c        instead of calculating effdeltaz constrain adjustment-vector-element
c        b(jj) directly:
c
         b(jj)=b(jj)-(e(4,i)-zmin)
c
         e(4,i)=zmin
         iconstrain(i)=1
c         write(16,'('' ***** depth constrained for event '',i5)') i
      endif
c
  29  continue  ! loop over all events (adjust)
c
c     all new hypocenters adjusted now.
c
c
c  do velocity adjustments here:
c
      if(scale(6).eq.0.0) goto 44  ! no velocity-adjustments...
      j1=4*neqs+nshot+1
      j2=j1+nltot-1
      k=0
      if(.not.single_turbo)then
         write(16,*)
         write(16,40)
 40      format(' doing velocity adjustments now...')
      endif
c
c     do it for all velocity-models
c
      do 4242 i=1,nmod
      k2=laysum(i)
      j11=j1+k2-1
      j22=j11+nplay(i)-1
      kj=1
c
c     do it for all layers in this model ( = i )
c
      do 42 jjj=j11,j22
      vp(i,kj)=vp(i,kj)+b(jjj)   !   <<---- VELOCITY-ADJUSTMENT
c
c     If reflected phases are used, the velocity of the layer just above
c     the reflector is NOT allowed to have a lower velocity than any of
c     the above layers!
c     Because usually the reflector is almost at the bottom of the model,
c     namely the MOHO, we simplify the condition and don't allow any
c     low-velocity-layers at all in the whole model (of course  o n l y
c     if reflected phases are used in this VELEST-run !)
      if(lowveloclay.eq.0)then
         if(kj.gt.1)then
            if(vp(i,kj).lt.vp(i,kj-1))then
               i f (.not.single_turbo)then
               write(16,'(1x,''WARNING: Tried to introduce a '',
     &                    ''low-velocity-layer! (Layer '',i2,'')'')') kj
               write(16,'(1x,''Setting DVP from '',f5.2,'' to 0.0 and'',
     &                    '' VP to vp(layer_above)+0.001'')') b(jjj)
               endif
               write(6,'(1x,''WARNING: Tried to introduce a '',
     &                    ''low-velocity-layer! (Layer '',i2,'')'')') kj
               write(6,'(1x,''Setting DVP from '',f5.2,'' to 0.0 and'',
     &                    '' VP to vp(layer_above)+0.001'')') b(jjj)
               write(6,*)
               b(jjj)=0.0
               vp(i,kj)=vp(i,kj-1)+0.001
            endif
         endif
      endif
c
      kj=kj+1
42    continue
c
4242  continue
c
c     each velocity-model adjusted now.
c
c     adjust p- & s- station-corrections here:
c
 44   if(scale(5).eq.0.0) goto 72 ! no station-correction adjustments...
      if(.not.single_turbo)then
         write(16,*)
         write(16,'('' doing station-correction adjustments...'')')
         write(16,*)
      endif
      k1=4*neqs+nshot+nltot+1
      ksta1=ksta
      if(nsp.eq.2) ksta1=ksta/2
      k2=k1+ksta1-1
c
c     do p-correction for all stations:
c
      do j=1,nsta
         cc(j)=0.
         if(map1(j).eq.0) goto 49
         if(map1(j).gt.ksta1) goto 49
         kk1=k1-1+map1(j)
         cc(j)=b(kk1)             !  STATION-CORRECTION-
         ptcor(j)=ptcor(j)+cc(j)  !  ADJUSTMENT !!! (p-correction)
 49      continue
      enddo
      if(nsp.ne.2) goto 72
      k1=4*neqs+nshot+nltot+1+ksta1
      ksta2=ksta-ksta1
      k2=k1+ksta2-1
c
c     do s-correction for all stations:
c
      do j=1,nsta
         cc(j)=0.
         if(map1(j).eq.0) goto 74
         if(map1(j).gt.ksta2) goto 74
         kk1=k1-1+map1(j)
         cc(j)=b(kk1)              ! STATION-CORRECTION-
         stcor(j)=stcor(j)+cc(j)   ! ADJUSTMENT !!! (s-correction)
74       continue
      enddo
c
72    continue
c
      RETURN
c
      end ! of subr. adjustmodel

File: src/sub/adjustmodel.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include "../include/globals.h"

extern int legs, neqs, nshot, nitt, isingle, itopo, ifixsolution;
extern int nsta, nsp, ksta, nltot, nmod, lowveloclay;
extern int knobs[IEQ], igap[IEQ], isconstrain[3], iconstrain[IEQ];
extern int map1[IST], iyr[IEQ], imo[IEQ], iday[IEQ], ihr[IEQ], imin[IEQ];
extern int nplay[INLTOT], laysum[INLTOT];

extern double zadj, zmin;
extern double e[5][IEQ], b[INVA * 2], pt[IST][IEQ], d[IST][3][IEQ];
extern double delta, scale[7], vp[ITOTMODELS][INLTOT], ptcor[IST], stcor[IST];
extern bool single_turbo;
extern FILE *fm_ptr;
extern char topo1file[81], topo2file[81];

extern int juliam(int iyr, int imo, int idy, int ihr, int imn);
extern void timeclear(int *iyr, int *imo, int *iday, int *ihr, int *imin,
                      double *sec, int *itime);
extern void chtop(double xx, double yy, double *zk, const char *topo1file,
                  const char *topo2file);

void adjustmodel(double damp) {
    (void)damp;

    int jj = -1;

    for (int i = 0; i < legs; ++i) {
        int n = (i >= neqs) ? 1 : 4;

        int year = (iyr[i] < 100) ? (iyr[i] + 1900) : iyr[i];
        int iminold = juliam(year, imo[i], iday[i], ihr[i], imin[i]);
        // Debug: print the shape of e
        fprintf(stderr, "DEBUG: e shape = [%zu][%zu]\n", sizeof(e)/sizeof(e[0]), sizeof(e[0])/sizeof(e[0][0]));
        fprintf(stderr, "DEBUG: i=%d, e[0]=%f, e[1]=%f, e[2]=%f, e[3]=%f, e[4]=%f\n", i, e[0][i], e[1][i], e[2][i], e[3][i], e[4][i]);
        for (int j = 0; j < n; ++j) {
            ++jj;
            e[j][i] = e[j][i] + b[jj];
        }

        if (e[0][i] < 0.0) {
            int itime = iminold;
            timeclear(&iyr[i], &imo[i], &iday[i], &ihr[i], &imin[i], &e[0][i], &itime);
            for (int j = 0; j < knobs[i] && j < IST; ++j) {
                pt[j][i] = pt[j][i] + (double)(iminold - itime) * 60.0;
            }
        }

        for (int j = 0; j < 3; ++j) {
            isconstrain[j] = 0;
        }
        iconstrain[i] = 0;

        if (isingle != 0) {
            if (nitt < 2) {
                isconstrain[0] = 1;
                e[3][i] = e[3][i] - b[3];
                b[3] = 0.0;
            }

            if (igap[0] > 250) {
                double dmin = 999.9f;
                for (int k = 0; k < knobs[0] && k < IST; ++k) {
                    delta = sqrtf((e[1][0] - d[k][0][0]) * (e[1][0] - d[k][0][0]) +
                                  (e[2][0] - d[k][1][0]) * (e[2][0] - d[k][1][0]));
                    if (delta < dmin) {
                        dmin = delta;
                    }
                }
                if (dmin > 15.0f) {
                    isconstrain[1] = 1;
                    if (iconstrain[i] == 1) {
                        iconstrain[i] = 3;
                    }
                    e[3][i] = e[3][i] - b[3];
                    b[3] = 0.0;
                }
            }

            if (fabs(b[3]) > zadj) {
                isconstrain[2] = 1;
                if (b[3] > 0.0) {
                    e[3][i] = e[3][i] - b[3] + zadj;
                    b[3] = zadj;
                } else {
                    e[3][i] = e[3][i] - b[3] - zadj;
                    b[3] = -zadj;
                }
            }
        } else {
            if (i < neqs && fabs(b[jj]) > zadj) {
                iconstrain[i] = 1;
                if (b[jj] > 0.0) {
                    e[3][i] = e[3][i] - b[jj] + zadj;
                    b[jj] = zadj;
                } else {
                    e[3][i] = e[3][i] - b[jj] - zadj;
                    b[jj] = -zadj;
                }
            }
        }

        if (itopo > 0) {
            if (e[3][i] < 0.0) {
                chtop(-e[1][i], e[2][i], &zmin, topo1file, topo2file);
            } else {
                zmin = 0.0;
            }
        }

        if (e[3][i] < zmin || (ifixsolution > 0 && e[3][i] <= 0.0)) {
            b[jj] = b[jj] - (e[3][i] - zmin);
            e[3][i] = zmin;
            iconstrain[i] = 1;
        }
    }

    if (scale[5] != 0.0) {
        if (!single_turbo) {
            FILE *out = fm_ptr ? fm_ptr : stdout;
            fprintf(out, "\n doing velocity adjustments now...\n");
        }

        int j1 = 4 * neqs + nshot; /* 0-based start */

        for (int i = 0; i < nmod; ++i) {
            int k2 = laysum[i];                 /* Fortran 1-based */
            int j11 = j1 + k2 - 1;              /* convert */
            int j22 = j11 + nplay[i] - 1;
            int kj = 1;

            for (int jjj = j11; jjj <= j22; ++jjj) {
                vp[i][kj - 1] = vp[i][kj - 1] + b[jjj];

                if (lowveloclay == 0 && kj > 1) {
                    if (vp[i][kj - 1] < vp[i][kj - 2]) {
                        if (!single_turbo) {
                            FILE *out = fm_ptr ? fm_ptr : stdout;
                            fprintf(out, " WARNING: Tried to introduce a low-velocity-layer! (Layer %2d)\n", kj);
                            fprintf(out, " Setting DVP from %5.2f to 0.0 and VP to vp(layer_above)+0.001\n", b[jjj]);
                        }
                        fprintf(stderr, " WARNING: Tried to introduce a low-velocity-layer! (Layer %2d)\n", kj);
                        fprintf(stderr, " Setting DVP from %5.2f to 0.0 and VP to vp(layer_above)+0.001\n", b[jjj]);
                        b[jjj] = 0.0;
                        vp[i][kj - 1] = vp[i][kj - 2] + 0.001f;
                    }
                }
                kj += 1;
            }
        }
    }

    if (scale[4] != 0.0) {
        if (!single_turbo) {
            FILE *out = fm_ptr ? fm_ptr : stdout;
            fprintf(out, "\n doing station-correction adjustments...\n\n");
        }

        int k1 = 4 * neqs + nshot + nltot + 1; /* Fortran 1-based */
        int ksta1 = ksta;
        if (nsp == 2) {
            ksta1 = ksta / 2;
        }

        for (int j = 0; j < nsta; ++j) {
            if (map1[j] == 0) continue;
            if (map1[j] > ksta1) continue;
            int kk1 = k1 - 1 + map1[j];
            int idx = kk1 - 1;
            ptcor[j] = ptcor[j] + b[idx];
        }

        if (nsp == 2) {
            k1 = 4 * neqs + nshot + nltot + 1 + ksta1;
            int ksta2 = ksta - ksta1;
            for (int j = 0; j < nsta; ++j) {
                if (map1[j] == 0) continue;
                if (map1[j] > ksta2) continue;
                int kk1 = k1 - 1 + map1[j];
                int idx = kk1 - 1;
                stcor[j] = stcor[j] + b[idx];
            }
        }
    }
}
